Better edit playlist logic

hh2010 opened the pull request:
addresses https://github.com/riverscuomo/boxify/issues/145

the communication between the UI and the database were not in sync. this cleans up the code and makes it so changes to playlist are immediately reflected locally, and tracked in the player state.

riverscuomo commented:
@hh2010 I'm not sure what I'm doing wrong but editing a playlist (resequencing or deleting tracks or renaming playlist) doesn't seem to trigger any logger statements I add to your new code blocks. And I think I'm still seeing some of the issues in #145 .
I definitely see your update code on my machine though.

hh2010 commented:
> @hh2010 okay, I'm running this branch in Examplify now, instead of Weezify, and I get this exception when I tap Edit Playlist:
> 
> ```
> 
> ════════ Exception caught by widgets library ═══════════════════════════════════
> The following assertion was thrown while finalizing the widget tree:
> Multiple widgets used the same GlobalKey.
> The key [_ReorderableListViewChildGlobalKey ValueKey<String>#53aa1] was used by multiple widgets. The parents of those widgets were:
> - Semantics(container: true, properties: SemanticsProperties, renderObject: RenderSemanticsAnnotations#035d2 relayoutBoundary=up26 NEEDS-LAYOUT NEEDS-PAINT)
> - Semantics(container: true, properties: SemanticsProperties, renderObject: RenderSemanticsAnnotations#b6f8e relayoutBoundary=up26 NEEDS-PAINT)
> A GlobalKey can only be specified on one widget at a time in the widget tree.
> 
> When the exception was thrown, this was the stack:
> #0      BuildOwner._debugVerifyGlobalKeyReservation.<anonymous closure>.<anonymous closure>.<anonymous closure> (package:flutter/src/widgets/framework.dart:3205:13)
> framework.dart:3205
> #1      _LinkedHashMapMixin.forEach (dart:collection-patch/compact_hash.dart:633:13)
> compact_hash.dart:633
> #2      BuildOwner._debugVerifyGlobalKeyReservation.<anonymous closure>.<anonymous closure> (package:flutter/src/widgets/framework.dart:3149:20)
> framework.dart:3149
> #3      _LinkedHashMapMixin.forEach (dart:collection-patch/compact_hash.dart:633:13)
> compact_hash.dart:633
> #4      BuildOwner._debugVerifyGlobalKeyReservation.<anonymous closure> (package:flutter/src/widgets/framework.dart:3144:36)
> framework.dart:3144
> #5      BuildOwner._debugVerifyGlobalKeyReservation (package:flutter/src/widgets/framework.dart:3213:6)
> framework.dart:3213
> #6      BuildOwner.finalizeTree.<anonymous closure> (package:flutter/src/widgets/framework.dart:3267:11)
> framework.dart:3267
> #7      BuildOwner.finalizeTree (package:flutter/src/widgets/framework.dart:3342:8)
> framework.dart:3342
> #8      WidgetsBinding.drawFrame (package:flutter/src/widgets/binding.dart:1169:19)
> binding.dart:1169
> #9      RendererBinding._handlePersistentFrameCallback (package:flutter/src/rendering/binding.dart:468:5)
> binding.dart:468
> #10     SchedulerBinding._invokeFrameCallback (package:flutter/src/scheduler/binding.dart:1397:15)
> binding.dart:1397
> #11     SchedulerBinding.handleDrawFrame (package:flutter/src/scheduler/binding.dart:1318:9)
> binding.dart:1318
> #12     SchedulerBinding._handleDrawFrame (package:flutter/src/scheduler/binding.dart:1176:5)
> binding.dart:1176
> #13     _invoke (dart:ui/hooks.dart:312:13)
> hooks.dart:312
> #14     PlatformDispatcher._drawFrame (dart:ui/platform_dispatcher.dart:419:5)
> platform_dispatcher.dart:419
> #15     _drawFrame (dart:ui/hooks.dart:283:31)
> ```

thank you for looking into
im not able to easily recreate this issue
it seems to be due to duplicate key values for a playlist track
this would happen if the uuid is not unique or possibly when its null
i looked through the [examplify firestore tracks database](https://console.firebase.google.com/u/0/project/boxify-dev-96c50/firestore/databases/-default-/data/~2Fbundles) to see if theres any null uuids and didnt see any
can you tell me which tracks are in the playlist and what operation you are doing on the playlist so i can try to recreate?

hh2010 commented:
> @hh2010 okay, I'm running this branch in Examplify now, instead of Weezify, and I get this exception when I tap Edit Playlist:
> 
> ```
> 
> ════════ Exception caught by widgets library ═══════════════════════════════════
> The following assertion was thrown while finalizing the widget tree:
> Multiple widgets used the same GlobalKey.
> The key [_ReorderableListViewChildGlobalKey ValueKey<String>#53aa1] was used by multiple widgets. The parents of those widgets were:
> - Semantics(container: true, properties: SemanticsProperties, renderObject: RenderSemanticsAnnotations#035d2 relayoutBoundary=up26 NEEDS-LAYOUT NEEDS-PAINT)
> - Semantics(container: true, properties: SemanticsProperties, renderObject: RenderSemanticsAnnotations#b6f8e relayoutBoundary=up26 NEEDS-PAINT)
> A GlobalKey can only be specified on one widget at a time in the widget tree.
> 
> When the exception was thrown, this was the stack:
> #0      BuildOwner._debugVerifyGlobalKeyReservation.<anonymous closure>.<anonymous closure>.<anonymous closure> (package:flutter/src/widgets/framework.dart:3205:13)
> framework.dart:3205
> #1      _LinkedHashMapMixin.forEach (dart:collection-patch/compact_hash.dart:633:13)
> compact_hash.dart:633
> #2      BuildOwner._debugVerifyGlobalKeyReservation.<anonymous closure>.<anonymous closure> (package:flutter/src/widgets/framework.dart:3149:20)
> framework.dart:3149
> #3      _LinkedHashMapMixin.forEach (dart:collection-patch/compact_hash.dart:633:13)
> compact_hash.dart:633
> #4      BuildOwner._debugVerifyGlobalKeyReservation.<anonymous closure> (package:flutter/src/widgets/framework.dart:3144:36)
> framework.dart:3144
> #5      BuildOwner._debugVerifyGlobalKeyReservation (package:flutter/src/widgets/framework.dart:3213:6)
> framework.dart:3213
> #6      BuildOwner.finalizeTree.<anonymous closure> (package:flutter/src/widgets/framework.dart:3267:11)
> framework.dart:3267
> #7      BuildOwner.finalizeTree (package:flutter/src/widgets/framework.dart:3342:8)
> framework.dart:3342
> #8      WidgetsBinding.drawFrame (package:flutter/src/widgets/binding.dart:1169:19)
> binding.dart:1169
> #9      RendererBinding._handlePersistentFrameCallback (package:flutter/src/rendering/binding.dart:468:5)
> binding.dart:468
> #10     SchedulerBinding._invokeFrameCallback (package:flutter/src/scheduler/binding.dart:1397:15)
> binding.dart:1397
> #11     SchedulerBinding.handleDrawFrame (package:flutter/src/scheduler/binding.dart:1318:9)
> binding.dart:1318
> #12     SchedulerBinding._handleDrawFrame (package:flutter/src/scheduler/binding.dart:1176:5)
> binding.dart:1176
> #13     _invoke (dart:ui/hooks.dart:312:13)
> hooks.dart:312
> #14     PlatformDispatcher._drawFrame (dart:ui/platform_dispatcher.dart:419:5)
> platform_dispatcher.dart:419
> #15     _drawFrame (dart:ui/hooks.dart:283:31)
> ```

ok this error happens when there are duplicate tracks in a playlist.
i have fixed it by  making the track list key unique to the index.
i also fixed a thing where the playlst was not immediately refreshing in the UI upon clicking save.
on my end it looks pretty good now.
